sklearn/svm/_liblinear.pyx

```diff
@@ -14,36 +14,71 @@ include "_liblinear.pxi"
 cnp.import_array()
 
 
-def train_wrap(X, cnp.ndarray[cnp.float64_t, ndim=1, mode='c'] Y,
-               bint is_sparse, int solver_type, double eps, double bias,
-               double C, cnp.ndarray[cnp.float64_t, ndim=1] class_weight,
-               int max_iter, unsigned random_seed, double epsilon,
-               cnp.ndarray[cnp.float64_t, ndim=1, mode='c'] sample_weight):
+def train_wrap(
+    object X,
+    cnp.float64_t[::1] Y,
+    bint is_sparse,
+    int solver_type,
+    double eps,
+    double bias,
+    double C,
+    cnp.float64_t[:] class_weight,
+    int max_iter,
+    unsigned random_seed,
+    double epsilon,
+    cnp.float64_t[::1] sample_weight
+):
     cdef parameter *param
     cdef problem *problem
     cdef model *model
     cdef char_const_ptr error_msg
     cdef int len_w
+    cdef cnp.float64_t[::1] x_data
+    cdef cnp.int32_t[::1] x_indices
+    cdef cnp.int32_t[::1] x_indptr
+    cdef cnp.float64_t[:, ::1] x_array
 
     if is_sparse:
+        x_data = X.data
+        x_indices = X.indices
+        x_indptr = X.indptr
         problem = csr_set_problem(
-                (<cnp.ndarray>X.data).data, X.dtype == np.float64,
-                (<cnp.ndarray[cnp.int32_t,   ndim=1, mode='c']>X.indices).data,
-                (<cnp.ndarray[cnp.int32_t,   ndim=1, mode='c']>X.indptr).data,
-                (<cnp.int32_t>X.shape[0]), (<cnp.int32_t>X.shape[1]),
-                (<cnp.int32_t>X.nnz), bias, sample_weight.data, Y.data)
+            <char*> &x_data[0],
+            X.dtype == np.float64,
+            <char*> &x_indices[0],
+            <char*> &x_indptr[0],
+            (<cnp.int32_t>X.shape[0]),
+            (<cnp.int32_t>X.shape[1]),
+            (<cnp.int32_t>X.nnz),
+            bias,
+            <char*> &sample_weight[0],
+            <char*> &Y[0]
+        )
     else:
+        x_array = X
```

Comment:
@jjerphan Could you kindly guide me at this point? Over here X can be a float i.e float32 when we consider the logistic regression model. There some of the tests will fail since I have specified the type as float64. If I try to instead change `cdef cnp.float64_t[:, ::1] x_array` to `cdef floating[:, ::1] x_array` it does not work and gives a compile error `Invalid use of fused types, type cannot be specialized`.

Reply:
Thank you very much!